compiler/src/lib.rs

```diff
@@ -22,59 +23,137 @@ mod r#type;
-                main_chunk,
+                &env.fqn,
+                chunk,
                 chunk_id,
-                &main_env.variables,
-                &mut bytecode,
                 ctx,
+                &mut bytecode,
                 &mut cp,
+                line_provider,
             );
-            write_exported(&mut cp, &mut bytecode)?;
-        }
-        bytecode.emit_u32(content.function_count() as u32);
-        for (chunk_id, env, chunk) in content.function_chunks(&it) {
-            let ctx = EmitterContext {
-                environment: env,
-                engine: link_engine,
-                captures: &captures,
-                chunk_id,
-                is_script: false,
-            };
-            let name = &env.fqn;
-            let signature_idx = cp.insert_string(name.to_string());
-            bytecode.emit_constant_ref(signature_idx);
-            compile_chunk(chunk, chunk_id, &env.variables, &mut bytecode, ctx, &mut cp);
         }
     }
 
     write(writer, &bytecode, &cp)
 }
 
+fn compile_chunk(
+    name: &Name,
+    chunk: &Chunk,
+    id: SourceId,
+    ctx: EmitterContext,
+    bytecode: &mut Bytecode,
+    cp: &mut ConstantPool,
+    line_provider: Option<&dyn SourceLineProvider>,
+) {
+    // emit the function's name
+    let signature_idx = cp.insert_string(name);
+    bytecode.emit_constant_ref(signature_idx);
+
+    // emits chunk's code attribute
+    let segments = compile_chunk_code(chunk, id, bytecode, ctx, cp);
+
+    bytecode.emit_byte(line_provider.map_or(0, |_| 1));
+
+    if let Some(line_provider) = line_provider {
+        let Some(content_id) = ctx.engine.get_original_content(id) else { return };
+        compile_line_mapping_attribute(segments, content_id, bytecode, line_provider);
+    }
+}
+
+fn compile_line_mapping_attribute(
+    positions: Vec<(usize, u32)>,
+    content_id: ContentId,
+    bytecode: &mut Bytecode,
+    line_provider: &dyn SourceLineProvider,
+) {
+    bytecode.emit_byte(MAPPINGS_ATTRIBUTE);
+    let mut mappings: Vec<(usize, u32)> = Vec::new();
+
+    let Some(((first_pos, first_ip), positions)) = positions.split_first() else {
+        bytecode.emit_u32(0);
+        return
+    };
+    let mut last_pos= *first_pos;
+    let mut last_ip = *first_ip;
+
+    let mut last_line = usize::MAX;
+    for (pos, instruction) in positions.iter().copied() {
+        if instruction > last_ip {
+            let line = line_provider.get_line(content_id, last_pos).unwrap();
+            if last_line != line {
+                mappings.push((line, last_ip));
+            }
+            last_line = line;
+            last_ip = instruction;
+            last_pos = pos;
+            continue;
+        }
+        last_pos = pos;
+    }
+
+    if mappings.is_empty() {
+        // if no mappings are set, bind first pos' line with first instruction.
+        let line = line_provider.get_line(content_id, *first_pos).unwrap();
+        mappings.push((line, 0))
+    }
```

Comment:
Given that this function has gradually increased its complexity, it could be better to split the responsibilities. Make that the `Instructions::push_position` method don't re-push on the same start segment to de-duplicate ahead of time.

Reply:
i'd like to let the `Instructions::push_position` as simple as possible and make all the computation if the mappings are required

Reply:
The tradeoff doesn't seem fair, as it increases code complexity here and increases memory usage as a lot is duplicated.